src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+                var referenceDocuments = PooledHashSet<DocumentId>.GetInstance();
+
+                foreach (var id in _state.DocumentIds)
+                {
+                    var result = await ChangeNamespaceToMatchFoldersAsync(solutionAfterNamespaceChange, id, cancellationToken).ConfigureAwait(false);
+                    solutionAfterNamespaceChange = result.Item1;
+                    referenceDocuments.AddRange(result.Item2);
+                }
+
+                var solutionAfterFirstMerge = await MergeDiffAsync(_state.Solution, solutionAfterNamespaceChange, cancellationToken).ConfigureAwait(false);
+
+                // After changing documents, we still need to remove unnecessary imports related to our change.
+                // We don't try to remove all imports that might become unnecessary/invalid after the namespace change, 
+                // just ones that fully matche the old or new namespace.
+                // For example, if we are changing namespace `Foo.Bar` to `A.B`, the using of name `Bar` below would remain 
+                // untouched:
+                //
+                //      namespace Foo
+                //      {
+                //          using Bar;
+                //          ~~~~~~~~~
+                //      }
+                //
+                // Also, because we may have added different imports to document that triggered the refactoring
+                // and the documents that reference affected types declared in changed namespace, we try to remove
+                // unnecessary imports separately.
+                var solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(
+                    solutionAfterFirstMerge,
+                    _state.DocumentIds, 
+                    CreateAllContainingNamespaces(OldNamespaceParts), 
+                    cancellationToken).ConfigureAwait(false);
+
+                solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(
+                    solutionAfterImportsRemoved, 
+                    referenceDocuments.ToImmutableArray(), 
+                    ImmutableArray.Create(_state.DeclaredNamespace, _state.TargetNamespace), 
+                    cancellationToken).ConfigureAwait(false);
+
+                var solutionAfterSecondMerge = await MergeDiffAsync(solutionAfterFirstMerge, solutionAfterImportsRemoved, cancellationToken).ConfigureAwait(false);
+
+                referenceDocuments.Free();
+                return ImmutableArray.Create<CodeActionOperation>(new ApplyChangesOperation(solutionAfterSecondMerge));
+            }
+
+            private ImmutableArray<string> CreateAllContainingNamespaces(ImmutableArray<string> parts)
+            {
+                var builder = ArrayBuilder<string>.GetInstance();
+                for (var i = 1; i <= OldNamespaceParts.Length; ++i)
+                {
+                    builder.Add(string.Join(".", OldNamespaceParts.Take(i)));
+                }
```

Comment:
is there a reason we generae all these different forms?  i'm guessing it's to give the user multiple options to invoke.  but that isn't clearly documented anywhere (afaict).  canthis be documented to explain the cases it is handling?

Reply:
There's comments at the callsite of this method. Given that what this method is doing is fairly obvious, I'd prefer to have the comment of why we need the result of this method at the callsite, instead of calling it out here. We need to add import for all containing namespaces (based on old declaration), in case there's reference to types in those namespace that are implicitly imported. 

---
In reply to: [230470164](https://github.com/dotnet/roslyn/pull/30156#discussion_r230470164) [](ancestors = 230470164)